### PR TITLE
Updated nvmecontrol wdc module to support log fetch operation for latest WDC NVMe products

### PR DESCRIPTION
**Pull request by AkhileshRN**

### CODE REVIEW

File: sbin/nvmecontrol/modules/wdc/wdc.c [hunk 1]
```diff
@@ -108,25 +129,26 @@ wdc_append_serial_name(int fd, char *buf, size_t len, const char *suffix)
...
 {
 	struct nvme_pt_command	pt;
 
 	memset(&pt, 0, sizeof(pt));
 	pt.cmd.opc = opcode;
-	pt.cmd.cdw10 = htole32(len / sizeof(uint32_t));	/* - 1 like all the others ??? */
-	pt.cmd.cdw11 = htole32(off / sizeof(uint32_t));
+	pt.cmd.cdw10 = htole32(len / sizeof(uint32_t));
 	pt.cmd.cdw12 = htole32(cmd);
+	if(e6lg_flag)
```
Comment (bsdimp) [inline sbin/nvmecontrol/modules/wdc/wdc.c:145]: This needs a space after the if.

File: sbin/nvmecontrol/modules/wdc/wdc.c [hunk 2]
```diff
@@ -135,17 +157,29 @@ wdc_get_data(int fd, uint32_t opcode, uint32_t len, uint32_t off, uint32_t cmd,
...
 	size_t resid;
+	bool e6lg_flag = false;
 
 	wdc_append_serial_name(fd, tmpl, MAXPATHLEN, suffix);
 
+	/* Read Log Dump header */
+	len = WDC_NVME_LOG_SIZE_HDR_LEN;
+	offset = 0;
+	hdr = malloc(len);
+	if(hdr == NULL)
```
Comment (bsdimp) [inline sbin/nvmecontrol/modules/wdc/wdc.c:176]: likewise

File: sbin/nvmecontrol/modules/wdc/wdc.c [hunk 3]
```diff
@@ -159,15 +193,13 @@ wdc_do_dump(int fd, char *tmpl, const char *suffix, uint32_t opcode,
...
 	do {
 		resid = len > NVME_MAX_XFER_SIZE ? NVME_MAX_XFER_SIZE : len;
-		wdc_get_data(fd, opcode, resid, offset, cmd, buf, resid);
+		wdc_get_data(fd, opcode, resid, offset, cmd, buf, resid, e6lg_flag);
 
 		if (first) {
 			len = be32dec(buf + len_off);
 			if (len == 0)
 				errx(1, "No data for %s", suffix);
-			if (memcmp("E6LG", buf, 4) != 0)
```
Comment (bsdimp) [inline sbin/nvmecontrol/modules/wdc/wdc.c:168]: why remove this sanity check for the data returned?

File: sbin/nvmecontrol/modules/wdc/wdc.c [hunk 4]
```diff
@@ -184,6 +216,144 @@ wdc_do_dump(int fd, char *tmpl, const char *suffix, uint32_t opcode,
...
+		err(1, "wdc_get_data_dui request failed");
+	if (nvme_completion_is_error(&pt.cpl))
+		errx(1, "wdc_get_data_dui request returned error");
+}
+
+static uint8_t
+wdc_get_dui_max_sections(uint16_t header_ver)
+{
+	switch (header_ver) {
+		case WDC_DUI_HEADER_VER_0:
```
Comment (bsdimp) [inline sbin/nvmecontrol/modules/wdc/wdc.c:250]: style nit: FreeBSD's style has these flush with the switch.

**Comment by bsdimp** [comment]
This code looks quite good.... I can easily fix the few style nits I saw and rebase this fix... But the one real concern was the removal of the sanity check to make sure we got the right header from the E6 logs. I'd snagged that from Linux and my past interactions with HGST/WD suggested that it was a good idea... 

And sorry for the delay... life's been interesting this year.

**Comment by bsdimp** [comment]
Also, this is a substantial change to the file. No one would blame you if you added a WD/HGST copyright to it.

**Comment by bsdimp** [comment]
Ah, I see the issue with E6LG... I hadn't noticed the test moved to accommodate other  blobs...
And this works for me for both of the HSGT/WDC drives I have handy (3.2TB SN100 and 1TB SN720).

I'll go ahead and commit this upstream and close this pull request. Should WDC want to pursue fixing the copyright, email imp@FreeBSD.org

**Comment by AkhileshRN** [comment]
> This code looks quite good.... I can easily fix the few style nits I saw and rebase this fix... But the one real concern was the removal of the sanity check to make sure we got the right header from the E6 logs. I'd snagged that from Linux and my past interactions with HGST/WD suggested that it was a good idea...
> 
> And sorry for the delay... life's been interesting this year. I'll give you a pass on the style issues and fix them myself as penance for being a little tardy on this.

Thanks!

**Comment by bsdimp** [comment]
I went ahead and committed the slightly cleaned up version w/o removing any of the sanity tests. If you could confirm it still works on the newer drives since I have only a limited number of variations. File a new pull request if there's any tweaks you discover this code still needs. I'll be trying it shortly on the newest WDC NVMe drives we're evaluating.

**Pull request closed without merging**